pruntime/enclave/src/contracts/diem.rs

```diff
@@ -0,0 +1,332 @@
+}
+
+impl Diem {
+    pub fn new() -> Self {
+        Diem {
+            accounts: Vec::new(),
+            transactions: BTreeMap::<String, Vec<Transaction>>::new(),
+            verified: BTreeMap::<String, bool>::new(),
+            seq_number: BTreeMap::<String, u64>::new(),
+        }
+    }
+
+    pub fn get_transaction(
+        &mut self,
+        transaction_with_proof: TransactionWithProof,
+        account_address: String,
+        address: AccountAddress,
+    ) -> Result<Transaction, Error> {
+        let transaction: Transaction = match bcs::from_bytes(&transaction_with_proof.transaction_bytes) {
+            Ok(tx) => tx,
+            Err(_) => {
+                println!("Decode transaction error");
+
+                return Err(Error::Other(String::from("Decode transaction error")));
+            }
+        };
+
+        if self.verified.get(&transaction.hash().to_hex()).is_some() {
+            return Ok(transaction);
+        }
+
+        let signed_tx: SignedTransaction = transaction.as_signed_user_txn().unwrap().clone();
+        if signed_tx.raw_txn.sender != address {
+            // Incoming tx doesn't need to be sequential
+            let mut found = false;
+            if let TransactionPayload::Script(script) = signed_tx.raw_txn.payload {
+                for arg in script.args {
+                    if let TransactionArgument::Address(recv_address) = arg {
+                        if recv_address == address {
+                            found = true;
+                            break;
+                        }
+                    }
+                }
+            }
+
+            if !found {
+                println!("Bad receiver address");
+
+                return Err(Error::Other(String::from("Bad receiver address")));
+            }
+        } else {
+            // Outgoing tx must be synced sequencely
+            if let Some(seq) = self.seq_number.get(&account_address) {
+                if seq + 1 != signed_tx.raw_txn.sequence_number {
+                    println!("Bad sequence number");
+
+                    return Err(Error::Other(String::from("Bad sequence number")));
+                }
+            }
+            self.seq_number.insert(account_address.clone(), signed_tx.raw_txn.sequence_number);
+        }
+
+        Ok(transaction)
+    }
+
+    pub fn verify_trusted_state(
+        &mut self,
+        transaction_with_proof: TransactionWithProof,
+    ) -> Result<(), Error> {
+        let epoch_change_proof: EpochChangeProof = transaction_with_proof.epoch_change_proof.clone();
+        let ledger_info_with_signatures: LedgerInfoWithSignatures = transaction_with_proof.ledger_info_with_signatures.clone();
+        let zero_ledger_info_with_sigs = epoch_change_proof.ledger_info_with_sigs[0].clone();
+        let mut trusted_state = TrustedState::try_from(zero_ledger_info_with_sigs.ledger_info())
+            .map_err(|_|Error::Other("get TrustedState error".to_string()))?;
```

Comment:
We should store `trusted_state` inside `struct Diem` contract, and update it by `new_state` we got in this function